Make WindowsServiceLifetime gracefully stop

ericstj opened the pull request:
WindowsServiceLifetime was not waiting for ServiceBase to stop the service.  As a result we would sometimes end the process before notifying service control manager that the service had stopped -- resulting in an error in the eventlog and sometimes a service restart.

We also were permitting multiple calls to Stop to occur - through SCM callbacks, and through public API.  We must not call SetServiceStatus again once the service is marked as stopped.

Fixes: #62579

ericstj committed:
Make WindowsServiceLifetime gracefully stop

WindowsServiceLifetime was not waiting for ServiceBase to stop the service.  As a result
we would sometimes end the process before notifying service control manager that the service
had stopped -- resulting in an error in the eventlog and sometimes a service restart.

We also were permitting multiple calls to Stop to occur - through SCM callbacks, and through
public API.  We must not call SetServiceStatus again once the service is marked as stopped.

ericstj commented on src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs:
```diff
@@ -647,6 +647,12 @@ public static void Run(ServiceBase service)
 
         public void Stop()
         {
+            if (_status.currentState == ServiceControlStatus.STATE_STOPPED || _status.currentState == default)
```
The docs for [SetServiceStatus](https://learn.microsoft.com/en-us/windows/win32/api/winsvc/nf-winsvc-setservicestatus) mention it must never be called after the service state is stopped:
> If the status is SERVICE_STOPPED, perform all necessary cleanup and call SetServiceStatus one time only. This function makes an LRPC call to the SCM. The first call to the function in the SERVICE_STOPPED state closes the RPC context handle and any subsequent calls can cause the process to crash.

We were actually hitting this in the common case, since we'd call back into the IHostLifetime StopAsync when SCM stopped the service.  Depending on how long it was since the first stop call we were seeing crashes (this was much more common when I inserted delay).  This check just prevents that re-entrancy.

It doesn't solve a race condition where Stop might be called from multiple threads (resulting in violating this constraint) or if an application itself called Stop at same time SCM was sending the stop message.  If we wanted to solve for these cases we'd need to introduce locking.

ericstj replied:
That latter case might actually be important -- since IHostLifetime.StopAsync gets called on a different thread and there is no guaranteed ordering that it happens after SCM's queued DeferredStop completes (that's the whole reason we had to wait in StopAsync), the two could be running concurrently.  I think we need better protection against this.

ericstj commented on src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs:
```diff
@@ -92,14 +93,14 @@ private void Run()
-            // Avoid deadlock where host waits for StopAsync before firing ApplicationStopped,
-            // and Stop waits for ApplicationStopped.
+            // Stop will cause the ServiceBase.Run method to complete and return, which completes _serviceStopped.
             Task.Run(Stop, CancellationToken.None);
```
This stop is redundant in the case that we received a stop signal from SCM.  It's here for the case where the app is deciding to stop the service itself.  We should try to avoid having this run concurrently with SCM's DeferredStop.

ericstj replied:
One thing that would be ordered is the call of OnStop.  We could set a flag at that point and honor here to avoid calling Stop if OnStop has already been called.

stephentoub reviewed (approved):
Is there any kind of test that can be added here?

ericstj commented:
> Is there any kind of test that can be added here?

Yeah but needs some work as we need a service project.  Plan to do so before moving out of draft.

ericstj committed:
Alternate approach to ensuring we only ever set STATE_STOPPED once.

ericstj committed:
Avoid calling ServiceBase.Stop on stopped service

I fixed double-calling STATE_STOPPED in ServiceBase, but this fix will
not be present on .NETFramework.  Workaround that by avoiding calling
ServiceBase.Stop when the service has already been stopped by SCM.

ericstj committed:
Add tests for WindowsServiceLifetime

These tests leverage RemoteExecutor to avoid creating a separate service
assembly.

stephentoub commented on src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs:
```diff
@@ -18,8 +18,10 @@ namespace Microsoft.Extensions.Hosting.WindowsServices
+        private readonly TaskCompletionSource<object?> _serviceStopped = new TaskCompletionSource<object?>(TaskCreationOptions.RunContinuationsAsynchronously);
         private readonly ManualResetEventSlim _delayStop = new ManualResetEventSlim();
         private readonly HostOptions _hostOptions;
+        private bool _isStopped;
```
The naming here between "isStopped" and "serviceStopped" is a little confusing.  What is the "is" referring to if not "service"?  I'm interpreting "isStopped" more as "stopRequested"?

ericstj replied:
I was trying to represent what SCM thought about the service state.  I was running into a conflict in name between the task `serviceStopped` and this flag (I even inadvertently named them the same thing at first).  I'm open to other naming suggestions that would be clearer.

Maybe for the task use `serviceRunComplete` or `serviceDispatcherStopped` then for the flag use `serviceStopped`?

ericstj replied:
I gave another try at the naming.  Will see if that's better.

stephentoub commented on src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs:
```diff
@@ -92,14 +94,18 @@ private void Run()
-            return Task.CompletedTask;
+            if (!_isStopped)
+            {
+                Task.Run(Stop, CancellationToken.None);
```
What do you want to have happen if Stop throws an exception?  Right now that exception is going to be eaten by the Task that's returned from Run and ignored (and at some point later the TaskSchedule.UnobservedTaskException event might be raised).

ericstj replied:
It was previously ignored, so I left it that way.  I suppose it wouldn't hurt to throw here?  
This is the body of that task.  It would throw if for some reason it failed to set the service state to stopped.
https://github.com/dotnet/runtime/blob/3f5cb1805ea040e0869c1da8b8a7476a223ce4d2/src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs#L502-L527
This is the caller:
https://github.com/dotnet/runtime/blob/3f5cb1805ea040e0869c1da8b8a7476a223ce4d2/src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs#L149

To capture the exception would you reccomend `return Task.Run(Stop, CancellationToken.None).ContinueWith(_serviceStopped.Task)`?

stephentoub replied:
It depends on what we want to have happen.

If Stop throws an exception, does that mean _serviceStopped will never complete?  If that were the case someone doing `await StopAsync()` would hang, because the Task they're awaiting would never complete.  If it's really the case that either both complete successfully or at most one fails, then I think we want to do this:
```C#
public async Task StopAsync(CancellationToken cancellationToken)
{
    if (!isStopped)
    {
        await Task.Run(Stop, CancellationToken.None).ConfigureAwait(false);
    }

    return await _serviceStopped.Task;
}
```

ericstj replied:
Got it.  This is where I landed as well after reding your comments.  Thanks for reminding me about `async` keyword to save my hackery with ContinueWith. 

I think I can actually force this path to throw as well, with user-code that overrides OnStop.  I'll write some tests for that.

ericstj commented on src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs:
```diff
@@ -92,14 +94,18 @@ private void Run()
             {
                 _delayStart.TrySetException(ex);
             }
+            _serviceStopped.TrySetResult(null);
```
Your other question about exceptions makes me wonder who would ever observe an exception thrown from this.  I think we could also propagate an exception if one occurred during stop - though it would likely be our own, since the only thing running on the dispatcher thread is ServiceBase.

stephentoub replied:
>  I think we could also propagate an exception if one occurred during stop - though it would likely be our own, since the only thing running on the dispatcher thread is ServiceBase.

That's another answer to the other question: if we catch all exceptions inside of Stop and use one to complete _serviceStopped, then the code in StopAsync can remain as is, as Stop itself would never fail and would instead propagate its exception to _serviceStopped.

ericstj replied:
I think I'll do this for correctness.  There's no reason to hide exceptions from framework code -- even if we don't expect them.  That just makes things harder to debug in the future.

ericstj commented on src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs:
```diff
@@ -92,14 +94,18 @@ private void Run()
+            _serviceStopped.TrySetResult(null);
         }
 
         public Task StopAsync(CancellationToken cancellationToken)
```
Yes it's part of the interface https://github.com/dotnet/runtime/blob/967250cc86b4c613015066b47b882e7424fdb590/src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/IHostLifetime.cs#L27
Funny enough, none of the other implementations of this interface actually do anything in StopAsync so the cancellationToken is likely there because we asked for it in FxDC ;).

While we can't cancel the call to Stop, we could check before running it (throwing OCE right before the call to Stop by passing in the token), we could also cancel waiting for Run to complete.  That doesn't really cancel the work since the thread would still be active running the native service dispatcher loop.  We cannot "rollback" anything once the `Stop` method has been called - that's a one way operation.

What's our thinking on this - what's the right meaning for cancel on a Stop operation?

stephentoub replied:
> we could check before running it (throwing OCE right before the call to Stop by passing in the token), we could also cancel waiting for Run to complete

We should do that upfront check.  We could also pass the token to Task.Run, which will cause that task to be canceled if cancellation is requested before the delegate has been invoked.  Canceling the wait on that Task.Run (e.g. via a WaitAsync) is a larger leap, as it means you end up having work that's pending out there that's become disconnected from the caller's Task and they have no way to track it at that point; we should avoid doing that unless we have a very good reason. Of course, if you're about to tear down the process anyway, maybe it doesn't matter.

ericstj replied:
Sounds good I made this change and added a test.

> Of course, if you're about to tear down the process anyway, maybe it doesn't matter.

I'm reluctant to do this.  Not allowing Run to exit gracefully means SCM may not treat the service as stopped, and if it's an auto-start service, it will *restart* it.

ericstj committed:
Respond to feedback and add more tests.

This better integrates with the RemoteExecutor component as well,
by hooking up the service process and fetching its handle.

This gives us the correct logging and exitcode handling from
RemoteExecutor.

ericstj commented:
These tests are failing in CI but passing locally:
https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-83892-merge-4425fb85bd7d4689a3/Microsoft.Extensions.Hosting.WindowsServices.Tests/1/console.f3be89b1.log?helixlogtype=result
I pulled the test payload and it's missing bindingRedirects for System.Runtime.CompilerServices.Unsafe 😕 I have those when building locally - emitted by RAR.  Might need to pull the binlogs from the test build to see what's going wrong on the build machine.

ericstj committed:
Honor Cancellation in StopAsync

ericstj commented:
So it looks like this difference is due to the following.

In CI builds we don't reference Microsoft.NET.Test.Sdk: 
https://github.com/dotnet/runtime/blob/ef6b6d581fbcdf65516d59c0597288755d46cb57/eng/testing/xunit/xunit.props#L26

The Microsoft.NET.Test.Sdk sets `GenerateBindingRedirectsOutputType` to true.
https://github.com/microsoft/vstest/blob/681849f4dced5de25b330cf9d20fe643de9e6924/src/package/Microsoft.NET.Test.Sdk/netfx/Microsoft.NET.Test.Sdk.targets#L22

Arcade sets the property, but does so after RAR, so it doesn't help (I have no idea why it's done after RAR):
https://github.com/dotnet/arcade/blob/747f53d751983dd062f39f4654bff074536e0012/src/Microsoft.DotNet.Arcade.Sdk/tools/Workarounds.targets#L6

So nothing is telling RAR to find and generate bindingRedirects in our test projects.

So that raises the question -- how are these loads succeeding for our test projects in helix today?  The answer is they aren't, initially:
```
LOG: Assembly download was successful. Attempting setup of file: C:\src\dotnet\runtime\artifacts\bin\Microsoft.Extensions.Hosting.WindowsServices.Tests\Debug\net462\System.Runtime.CompilerServices.Unsafe.dll
LOG: Entering run-from-source setup phase.
LOG: Assembly Name is: System.Runtime.CompilerServices.Unsafe, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
WRN: Comparing the assembly name resulted in the mismatch: Major Version
ERR: The assembly reference did not match the assembly definition found.
```

What happens is Xunit has installed an AssemblyResolve handler that probes the [test directory](https://github.com/xunit/xunit/blob/1892224c60f8499b42ee1d8a521d65995ef32aea/src/xunit.runner.utility/Frameworks/v2/Xunit2Discoverer.cs#L95) and ignores assembly version completely:
https://github.com/xunit/xunit/blob/1892224c60f8499b42ee1d8a521d65995ef32aea/src/common/AssemblyResolution/AssemblyHelper_Desktop.cs#L53

So this will break anything in remote-executor scenarios that requires bindingRedirects, since they won't be using that runner to set up the AssemblyResolve handler -- they need the bindingRedirects to be correct.

ericstj committed:
Fix bindingRedirects in RemoteExecutor

ericstj committed:
Use Async lambdas for service testing

ericstj committed:
Fix issue on Win7 where duplicate service descriptions are disallowed

ericstj commented:
This should fix the tests 🤞

ericstj commented:
@buyaa-n can I get your review of this?

buyaa-n commented on src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs:
```diff
@@ -18,8 +18,10 @@ namespace Microsoft.Extensions.Hosting.WindowsServices
+        private readonly TaskCompletionSource<object?> _serviceDispatcherStopped = new TaskCompletionSource<object?>(TaskCreationOptions.RunContinuationsAsynchronously);
         private readonly ManualResetEventSlim _delayStop = new ManualResetEventSlim();
         private readonly HostOptions _hostOptions;
+        private bool _serviceStopped;
```
NIT: it looks to me like used for checking if `_stopRequested`

ericstj replied:
@buyaa-n was a naming suggestion?

buyaa-n replied:
Yeah, if it makes sense for you

ericstj replied:
I'll make it `_serviceStopRequested` -- though at this point it's irreversible.  I tried to fix up some of the public docs to reflect this.

buyaa-n reviewed (approved):
LGTM, thanks for the broad testing!

ericstj commented on src/libraries/Microsoft.Extensions.Hosting.WindowsServices/tests/WindowsServiceLifetimeTests.cs:
```diff
@@ -0,0 +1,338 @@
+
+            serviceTester.WaitForStatus(ServiceControllerStatus.Stopped);
+            var status = serviceTester.QueryServiceStatus();
+            Assert.Equal(1064, status.win32ExitCode);
```
Yeah, I can use the errors from https://learn.microsoft.com/en-us/windows/win32/debug/system-error-codes--1000-1299-  good suggestion.

ericstj commented on src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs:
```diff
@@ -116,6 +125,7 @@ protected override void OnStart(string[] args)
         /// <remarks>This might be called multiple times by service Stop, ApplicationStopping, and StopAsync. That's okay because StopApplication uses a CancellationTokenSource and prevents any recursion.</remarks>
```
This comment is now wrong and in fact was a signal that there was an issue here.

I fixed the extra call from `StopAsync`, the call from `ApplicationStopping` existed at [the time of this comment](https://github.com/dotnet/extensions/commit/05d1a6eb0e051a0cbadc64246f77fe9b836d33a6#diff-e4d0d3a9aa655704afc92739fc3402911ce12676b9e316d14f45cbc16f3fca4eR38) but was removed [later](https://github.com/dotnet/extensions/commit/43435a80d007d568aa17d891ea32ca87363716fc#diff-899fddd5f7ff286e68af676af2b391e6e284d41a2c02ed3e1678c997dd843e06L38).  At this point `OnStop` should only ever be called once.

ericstj committed:
Respond to feedback

buyaa-n commented on src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs:
```diff
@@ -87,19 +89,30 @@ private void Run()
+                await Task.Run(Stop, cancellationToken).ConfigureAwait(false);
+            }
+
+            // When the underlying service is stopped this will cause the ServiceBase.Run method to complete and return, which completes _serviceStopRequested.
```
NIT: Seems it is about `_serviceDispatcherStopped`
```suggestion
            // When the underlying service is stopped this will cause the ServiceBase.Run method to complete and return, which completes _serviceDispatcherStopped.
```

ericstj committed:
Fix comment and add timeout

ericstj commented:
JIT test failure is https://github.com/dotnet/runtime/issues/84398.

ericstj commented:
/backport to release/7.0-staging

ericstj commented:
/backport to release/6.0-staging
